lib/internal/main/watch_mode.js

```diff
@@ -0,0 +1,122 @@
+
+const { spawn } = require('child_process');
+const { watch } = require('fs/promises');
+const { inspect } = require('util');
+const { setTimeout, clearTimeout } = require('timers');
+const { dirname, sep, resolve } = require('path');
+const { once } = require('events');
+
+
+prepareMainThreadExecution(false);
+markBootstrapComplete();
+
+// TODO(MoLow): Make kill signal configurable
+const kKillSignal = 'SIGTERM';
+const kWatchedPaths = ArrayPrototypeMap(getOptionValue('--watch-path'), (path) => resolve(path));
+const kCommand = ArrayPrototypeSlice(process.argv, 1);
+const kCommandStr = inspect(ArrayPrototypeJoin(kCommand, ' '));
+const args = ArrayPrototypeReduce(process.execArgv, (acc, flag, i, arr) => {
+  if (arr[i] !== '--watch-path' && arr[i - 1] !== '--watch-path' && arr[i] !== '--watch') {
+    acc.push(arr[i]);
+  }
+  return acc;
+}, []);
+ArrayPrototypePushApply(args, kCommand);
+
+function isWatchedFile(filename) {
+  if (kWatchedPaths.length > 0) {
+    return ArrayPrototypeSome(kWatchedPaths, (path) => StringPrototypeStartsWith(filename, path));
+  }
+
+  const directory = dirname(filename);
+  if (directory === '.') {
+    return true;
+  }
+
+  const dirs = StringPrototypeSplit(directory, sep);
+  return !ArrayPrototypeSome(dirs, (dir) => dir[0] === '.' || dir === 'node_modules');
+}
+
+function debounce(fn, duration = 100) {
+  let timeout;
+  return () => {
+    if (timeout) {
+      clearTimeout(timeout);
+    }
+
+    timeout = setTimeout(fn, duration).unref();
+  };
+}
+
+function exitHandler(code) {
+  if (code === 0) {
+    process.stdout.write(`${blue}Completed running ${kCommandStr}${white}\n`);
+  } else {
+    process.stdout.write(`${red}Failed running ${kCommandStr}${white}\n`);
+  }
+}
+
+let graceTimer;
+function reportGracefulTermination() {
+  let reported = false;
+  clearTimeout(graceTimer);
+  graceTimer = setTimeout(() => {
+    reported = true;
+    process.stdout.write(`${blue}Waiting for graceful termination${white}\n`);
+  }, 1000).unref();
+  return () => {
+    clearTimeout(graceTimer);
+    if (reported) {
+      process.stdout.write(`${green}gracefully terminated${white}\n`);
+    }
+  };
+}
+
+
+let childProcess;
+async function run(restarting) {
+  if (restarting) {
+    process.stdout.write('\u001Bc');
+    process.stdout.write(`${green}Restarting ${kCommandStr}${white}\n`);
+  }
+  if (childProcess && !childProcess.killed) {
+    childProcess.removeListener('exit', exitHandler);
+    const onExit = once(childProcess, 'exit');
+    const clearGraceReport = reportGracefulTermination();
+    childProcess.kill(kKillSignal);
+    await onExit;
+    clearGraceReport();
+  }
+
+  childProcess = spawn(process.execPath, args, { stdio: ['inherit', 'inherit', 'inherit'] });
+  childProcess.once('exit', exitHandler);
+}
+
+const restart = debounce(() => run(true));
+
+(async () => {
+  run();
+  const watcher = watch(process.cwd(), { recursive: true });
```

Comment:
I think it makes sense for `watch` to watch loaded modules - that is, if I `require` or `import` something it's watched otherwise it's not.

Reply:
I think there is no optimal solution here; watching one or more entire directories will likely lead to false positives (and perhaps false negatives), watching only loaded modules will lead to false negatives.

This probably needs to be configurable in more ways.

Reply:
I agree though I do think it's better DX to watch the loaded modules (like newer tools like node-dev/ts-node-dev) and not the folder (like older tools like nodemon) and allow watching folders by passing --watch-path

Reply:
I have added options for both watching loaded modules (default) or watching paths using `--watch-path`